### ISSUE TITLE
Druid Docker-Image not compatible with OpenShift default SecurityContextConstraint (SCC)

### ISSUE THREAD

**Issue by mstein11** [apache/druid#10777]
### Affected Version

Druid Version:
0.20.0

OpenShift Version:
4

### Description

When trying to start Druid in an OpenShift environment based on the [Helm-Chart](https://github.com/helm/charts/tree/master/incubator/druid), it fails. The Deployments of Druid-Broker, Druid-Coordinator, Druid-Router and Druid-Historical all fail start start their pods because of missing permission to create files at the container start. We assume this is because OpenShift does not run the pods as User 1000 but rather as a random UserId and the random User is not permitted to manipulate some files.

We further assume, that owning the required files by the root group (group Id 0) and making those files writeable by the group will alleviate the issue - as described [here](https://stackoverflow.com/questions/42363105/permission-denied-mkdir-in-container-on-openshift) and [here](https://docs.openshift.com/container-platform/3.3/creating_images/guidelines.html#openshift-container-platform-specific-guidelines)

We use the default helm values to deploy the druid to openshift.

### pod logs
Druid-Broker
2021-01-19T08:01:06+0000 startup service broker
Setting druid.host=10.86.0.214 in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.indexer.logs.type=file in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.extensions.loadList=["druid-histogram", "druid-datasketches", "druid-lookups-cached-global", "postgresql-metadata-storage"] in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.processing.numMergeBuffers=2 in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.emitter.http.recipientBaseUrl=http://druid_exporter_url:druid_exporter_port/druid in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.metadata.storage.type=postgresql in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.processing.numThreads=1 in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.metadata.storage.connector.connectURI=jdbc:postgresql://test-druid-postgresql:5432/druid in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.storage.type=local in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.metadata.storage.connector.user=druid in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.zk.service.host=test-druid-zookeeper-headless:2181 in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.indexer.logs.directory=/opt/data/indexing-logs in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.emitter=noop in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.metadata.storage.connector.password=druid in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.processing.buffer.sizeBytes=50000000 in /tmp/conf/druid/cluster/query/broker/runtime.properties
Setting druid.emitter.logging.logLevel=debug in /tmp/conf/druid/cluster/query/broker/runtime.properties
mkdir: can't create directory 'var/tmp': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
 
Coordinator:
2021-01-19T08:01:57+0000 startup service coordinator
Setting druid.host=10.86.0.215 in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.indexer.logs.type=file in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.extensions.loadList=["druid-histogram", "druid-datasketches", "druid-lookups-cached-global", "postgresql-metadata-storage"] in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.metadata.storage.type=postgresql in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.emitter.http.recipientBaseUrl=http://druid_exporter_url:druid_exporter_port/druid in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.metadata.storage.connector.connectURI=jdbc:postgresql://test-druid-postgresql:5432/druid in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.storage.type=local in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.metadata.storage.connector.user=druid in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.indexer.logs.directory=/opt/data/indexing-logs in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.zk.service.host=test-druid-zookeeper-headless:2181 in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.emitter=noop in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.metadata.storage.connector.password=druid in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
Setting druid.emitter.logging.logLevel=debug in /tmp/conf/druid/cluster/master/coordinator-overlord/runtime.properties
mkdir: can't create directory 'var/tmp': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
 
Router
2021-01-19T08:03:26+0000 startup service router
Setting druid.host=10.86.0.216 in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.indexer.logs.type=file in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.extensions.loadList=["druid-histogram", "druid-datasketches", "druid-lookups-cached-global", "postgresql-metadata-storage"] in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.emitter.http.recipientBaseUrl=http://druid_exporter_url:druid_exporter_port/druid in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.metadata.storage.type=postgresql in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.metadata.storage.connector.connectURI=jdbc:postgresql://test-druid-postgresql:5432/druid in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.storage.type=local in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.metadata.storage.connector.user=druid in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.indexer.logs.directory=/opt/data/indexing-logs in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.zk.service.host=test-druid-zookeeper-headless:2181 in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.emitter=noop in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.metadata.storage.connector.password=druid in /tmp/conf/druid/cluster/query/router/runtime.properties
Setting druid.emitter.logging.logLevel=debug in /tmp/conf/druid/cluster/query/router/runtime.properties
mkdir: can't create directory 'var/tmp': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied

Historical
2021-01-19T08:04:24+0000 startup service historical
Setting druid.host=10.86.0.221 in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.indexer.logs.type=file in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.extensions.loadList=["druid-histogram", "druid-datasketches", "druid-lookups-cached-global", "postgresql-metadata-storage"] in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.processing.numMergeBuffers=2 in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.emitter.http.recipientBaseUrl=http://druid_exporter_url:druid_exporter_port/druid in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.metadata.storage.type=postgresql in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.processing.numThreads=1 in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.metadata.storage.connector.connectURI=jdbc:postgresql://test-druid-postgresql:5432/druid in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.storage.type=local in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.metadata.storage.connector.user=druid in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.zk.service.host=test-druid-zookeeper-headless:2181 in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.indexer.logs.directory=/opt/data/indexing-logs in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.emitter=noop in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.processing.buffer.sizeBytes=50000000 in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.metadata.storage.connector.password=druid in /tmp/conf/druid/cluster/data/historical/runtime.properties
Setting druid.emitter.logging.logLevel=debug in /tmp/conf/druid/cluster/data/historical/runtime.properties
mkdir: can't create directory 'var/tmp': Permission denied

**Comment by utsavjha** [comment]
Hi @mstein11 , you are indeed correct. The UID alterations need to be made by the startup/entrypoint inside the /etc/passwd file. This is because the containers are run without a hostname in OC by default. 
I also noticed this issue while trying to use the `parquet-extensions` extension. The absence of hostname leads to failure in initialisation of Parquet based services inside druid. 

I used the following script to alter my `/etc/passwd` file and then set it to defaullt UID of root user, if incase it helps:

```
myuid=$(id -u)
mygid=$(id -g)
uidentry=$(getent passwd $myuid)

echo "Attempting to write to /etc/passwd since UIDEntry: $uidentry, UID: $myuid, GID: $mygid"
if [ -z "$uidentry" ] ; then
    # assumes /etc/passwd has root-group (gid 0) ownership
     cat /etc/passwd | sed -e "s/^$myuid:/builder:/" > /tmp/passwd
     echo "$myuid:x:$myuid:$mygid:,,,:/home/$myuid:/bin/bash" >> /tmp/passwd
     cat /tmp/passwd >> /etc/passwd
     rm /tmp/passwd
    echo "User Info added to /etc/passwd"
fi

echo "Locking further alterations"
chmod 400 /etc/passwd
```

**Comment by FrankChen021** [comment]
It seems like this problem was fixed by https://github.com/helm/charts/pull/23561 . If it still exists, please re-open the issue to let us know.

**Comment by hosswald** [comment]
@mstein11 @FrankChen021 @utsavjha I'm still experiencing exactly this problem with the helm chart at https://github.com/apache/druid/tree/master/helm/druid (6b0b1d7af317d1a6b219cf1213139abf1607eaa0).
It has securityContext.mountPath: /opt/druid/var/druid/ and fsGroup: 1000, but I'm getting:
CreateContainerConfigError `Error: container has runAsNonRoot and image has non-numeric user (druid), cannot verify user is non-root` for historical and middleManager, and logs
```
mkdir: can't create directory 'var/tmp': Permission denied
mkdir: can't create directory 'var/druid/': Permission denied
```
for the other services.
If I override the securityContext in my own values.yaml
```
druid:
  historical:
    securityContext: false
  middleManager:
    securityContext: false
```
historical and middleManager start, but have the Permission denied error as well.

**Comment by FrankChen021** [comment]
@hosswald I'm not an expert of helm. 
Could you try to change the mountPath from `/opt/druid/var/druid/` to `/opt/druid/var` to see if it works?

**Comment by mstein11** [comment]
We build an own image of druid with the following dockerfile: 

```
FROM apache/druid:0.22.1 as builder
 
FROM ubi8-openjdk-8:1.3
 
COPY --chown=1000:0 --from=builder /opt /opt
COPY --from=builder /druid.sh /druid.sh
 
USER root
RUN dnf install procps -y
RUN dnf install iproute -y
RUN dnf install less -y
USER 1000
 
VOLUME /opt/druid/var
WORKDIR /opt/druid
 
ENTRYPOINT ["/druid.sh"]
```

Hope this helps!

**Comment by hosswald** [comment]
@mstein11 much appreciated. However, I think I found an easier way (for me), see below.
@FrankChen021 that did not help. The problem is, OpenShift mounts the volume as owned by root:root.
To get around that, I had to convince OpenShift to let the container run with the user defined in the Dockerfile:
`oc adm policy add-scc-to-user anyuid -z default kubeadmin`
and use an init container to set the permissions in druid/templates/historical/statefulset.yaml (etc.):
```
      initContainers:
        - name: initcontainer
          image: alpine:latest
          command: [ "/bin/sh", "-c" ]
          args:
            - chown 1000:1000 /opt/druid/var/druid;
          volumeMounts:
            - mountPath: /opt/druid/var/druid/
              name: data
```
Result:
```
PS C:\Users\...\helm\my-project> oc debug my-project-druid-historical-0 -- ls -dl /opt/druid/var/druid
Defaulting container name to druid.
Use 'oc describe pod/my-project-druid-historical-0-debug -n my-project' to see all of the containers in this pod.

Debugging with pod/my-project-druid-historical-0-debug, original command: <image entrypoint>
Waiting for pod to start ...
drwxrwx---    7 druid    druid           94 Aug  9 12:20 /opt/druid/var/druid

Removing debug pod ...

```

**Comment by FrankChen021** [comment]
> The problem is, OpenShift mounts the volume as owned by root:root.

We used to encountered such problem on macOS, and we have fixed that problem in #11167

Seems like that this permission problem also exists on OpenShift. To eliminate such problem, maybe we can run druid as root user in the docker.
